src/hotspot/share/gc/serial/serialBlockOffsetTable.cpp

```diff
@@ -85,384 +81,109 @@ void BlockOffsetSharedArray::resize(size_t new_word_size) {
-    // of the next.
-    size_t reach = start_card - 1 + (BOTConstants::power_to_cards_back(i+1) - 1);
-    offset = BOTConstants::card_size_in_words() + i;
-    if (reach >= end_card) {
-      _array->set_offset_array(start_card_for_region, end_card, offset, reducing);
+// Write the backskip value for each region.
+//
+//    offset
+//    card             2nd                       3rd
+//     | +- 1st        |                         |
+//     v v             v                         v
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+     +-+-+-+-+-+-+-+-+-+-+-
+//    |x|0|0|0|0|0|0|0|1|1|1|1|1|1| ... |1|1|1|1|2|2|2|2|2|2| ...
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+     +-+-+-+-+-+-+-+-+-+-+-
+//    11              19                        75
+//      12
+//
+//    offset card is the card that points to the start of an object
+//      x - offset value of offset card
+//    1st - start of first logarithmic region
+//      0 corresponds to logarithmic value N_words + 0 and 2**(3 * 0) = 1
+//    2nd - start of second logarithmic region
+//      1 corresponds to logarithmic value N_words + 1 and 2**(3 * 1) = 8
+//    3rd - start of third logarithmic region
+//      2 corresponds to logarithmic value N_words + 2 and 2**(3 * 2) = 64
+//
+//    integer below the block offset entry is an example of
+//    the index of the entry
+//
+//    Given an address,
+//      Find the index for the address
+//      Find the block offset table entry
+//      Convert the entry to a back slide
+//        (e.g., with today's, offset = 0x81 =>
+//          back slip = 2**(3*(0x81 - N_words)) = 2**3) = 8
+//      Move back N (e.g., 8) entries and repeat with the
+//        value of the new entry
+//
+void BlockOffsetTable::update_for_block_work(HeapWord* blk_start,
+                                             HeapWord* blk_end) {
+  HeapWord* const cur_card_boundary = align_up_by_card_size(blk_start);
+  size_t const offset_card = _array->index_for(cur_card_boundary);
+
+  // The first card holds the actual offset.
+  _array->set_offset_array(offset_card, cur_card_boundary, blk_start);
+
+  // Check if this block spans over other cards.
+  size_t end_card = _array->index_for(blk_end - 1);
+  assert(offset_card <= end_card, "inv");
+
+  if (offset_card != end_card) {
+    // Handling remaining cards.
+    size_t start_card_for_region = offset_card + 1;
+    for (uint i = 0; i < BOTConstants::N_powers; i++) {
+      // -1 so that the reach ends in this region and not at the start
+      // of the next.
+      size_t reach = offset_card + BOTConstants::power_to_cards_back(i + 1) - 1;
+      u_char value = checked_cast<u_char>(BOTConstants::card_size_in_words() + i);
+      if (reach >= end_card) {
+        _array->set_offset_array(start_card_for_region, end_card, value);
+        start_card_for_region = reach + 1;
+        break;
+      }
+      _array->set_offset_array(start_card_for_region, reach, value);
       start_card_for_region = reach + 1;
```

Comment:
```suggestion
      _array->set_offset_array(start_card_for_region, MIN2(end_card, reach), value);
      start_card_for_region = reach + 1;
      if (reach >= end_card) {
        // Set all the cards covered by this region.
        break;
      }
```

I would probably try to merge these two paths as much as possible as suggested above (untested).